js/modules/loki_snode_api.js

```diff
@@ -0,0 +1,170 @@
+/* global log, window, Whisper */
+
+const fetch = require('node-fetch');
+const is = require('@sindresorhus/is');
+const dns = require('dns');
+
+// Will be raised (to 3?) when we get more nodes
+const MINIMUM_SWARM_NODES = 1;
+
+class LokiSnodeAPI {
+
+  constructor({ url, swarmServerPort }) {
+    if (!is.string(url)) {
+      throw new Error('WebAPI.initialize: Invalid server url');
+    }
+    this.url = url;
+    this.swarmServerPort = swarmServerPort
+      ? `:${swarmServerPort}`
+      : '';
+    this.swarmsPendingReplenish = {};
+    this.ourSwarmNodes = {};
+  }
+
+  getRandomSnodeAddress() {
+    /* resolve random snode */
+    return new Promise((resolve, reject) => {
+      dns.resolveCname(this.url, (err, address) => {
+        if(err) {
+          reject(err);
+        } else {
+          resolve(address[0]);
+        }
+      });
+    });
+  }
+
+  unreachableNode(pubKey, nodeUrl) {
+    if (pubKey === window.textsecure.storage.user.getNumber()) {
+      delete this.ourSwarmNodes[nodeUrl];
+    }
+  }
+
+  updateLastHash(nodeUrl, hash) {
+    if (!this.ourSwarmNodes[nodeUrl]) {
+      this.ourSwarmNodes[nodeUrl] = {
+        lastHash: hash,
+      }
+    } else {
+      this.ourSwarmNodes[nodeUrl].lastHash = hash;
+    }
+  }
+
+  async getOurSwarmNodes() {
+    if (
+      !this.ourSwarmNodes ||
+      Object.keys(this.ourSwarmNodes).length < MINIMUM_SWARM_NODES
+    ) {
+      // Try refresh our swarm list once
+      const ourKey = window.textsecure.storage.user.getNumber();
+      const nodeAddresses = await window.LokiSnodeAPI.getSwarmNodes(ourKey);
+
+      this.ourSwarmNodes = {};
+      nodeAddresses.forEach(url => {
+        this.ourSwarmNodes[url] = {};
+      })
+      if (!this.ourSwarmNodes || Object.keys(this.ourSwarmNodes).length === 0) {
+        throw Error('Could not load our swarm')
+      }
+    }
+    return this.ourSwarmNodes;
+  }
+
+  async getSwarmNodesByPubkey(pubKey) {
+    const swarmNodes = await window.Signal.Data.getSwarmNodesByPubkey(pubKey);
+    // TODO: Check if swarm list is below a threshold rather than empty
+    if (swarmNodes && swarmNodes.length !== 0) {
+      return swarmNodes;
+    }
+    return this.replenishSwarm(pubKey);
+  }
+
+  async replenishSwarm(pubKey) {
+    const conversation = window.ConversationController.get(pubKey);
+    if (!(pubKey in this.swarmsPendingReplenish)) {
+      this.swarmsPendingReplenish[pubKey] = new Promise(async (resolve) => {
+        const newSwarmNodes = await this.getSwarmNodes(pubKey);
```

Comment:
should we `catch` any exceptions (e.g. connectivity issues) here and return an empty array ?